### ISSUE TITLE
NotImplementedError for maximizing a convex function with regularization.

### ISSUE THREAD

**Issue by kosmitive** [cvxgrp/dccp#54]
When i try to solve a problem of the form with Q positive definite.

min x.T @ x - x.T @ Q @ x
s.t. x_min <= x <= x_max

using the DCCP method. When I plug this directly
into the solver it get an error saying not DCCP. I then
took a look at the paper and transformed the problem 
as

min x.T @ x - z
s.t. x_min <= x <= x_max
      x.T @ Q @ x == z

but then I get a NotImplementedError. Is there something
missing in the DCCP library? Or is the problem simply 
not solvable by this solver. Because as soon as I remove the regularization
term x.T @ x from the objective in the first problem, it can be solved.
Do you have a suggestion what's the problem?

**Comment by xinyueshen** [comment]
Hi @kosmitive, DCCP currently works with CVXPY 1.0 but not CVXPY 1.1. `@` is the multiplication operator in CVXPY 1.1, so you may want to switch back to CVXPY 1.0 and use `*` instead.

**Comment by kosmitive** [comment]
Hi @xinuyueshen, currently we are using the version 1.0.31 and switched to * sign, but the same error occurs. I just used these operators for notation, we are working over time with linear-constraints on subsequent x_t, but this problem would resemble the core idea. Please note that in some sense the solutions are at the boundary of the space X, because we want to maximize g(x)=x.T @ Q @ x and additionally regularize the x so that they don't jump from x_min to x_max over time, which is required in our case. So if we are using this fragment

 # build the problem
 prob = cp.Problem(cp.Minimize(reg - obj), constraints)
 prob.solve(method="dccp")

we'll get: "Exception: Problem is not DCCP". However if we transform using this

# convert to quadratic constrain
t_var = cp.Variable(1)
constraints.append(obj == t_var)

# build the problem
prob = cp.Problem(cp.Minimize(reg - t_var), constraints)
prob.solve(method="dccp")

which is from section C in https://web.stanford.edu/~boyd/papers/pdf/dccp_cdc.pdf. And when using the second fragment we'll receive "TypeError: cannot unpack non-iterable NotImplementedType object" after the following two AttributeErrors:

/home/kosmitive/miniconda3/envs/experiments/lib/python3.8/site-packages/cvxpy/atoms/quad_form.py:202: UserWarning: Forming a nonconvex expression quad_form(x, indefinite).
  warnings.warn("Forming a nonconvex expression quad_form(x, indefinite).")
Traceback (most recent call last):
  File "/home/kosmitive/miniconda3/envs/experiments/lib/python3.8/site-packages/cvxpy/utilities/performance_utils.py", line 27, in _lazyprop
    return getattr(self, attr_name)
AttributeError: 'AddExpression' object has no attribute '_lazy_canonical_form'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kosmitive/miniconda3/envs/experiments/lib/python3.8/site-packages/cvxpy/utilities/performance_utils.py", line 27, in _lazyprop
    return getattr(self, attr_name)
AttributeError: 'real' object has no attribute '_lazy_canonical_form'

But with 

# build the problem
prob = cp.Problem(cp.Minimize(-obj), constraints)
prob.solve(method="dccp")

it gets a solution. obj and reg are convex in x. Maybe this problem is not suitable for this solver and more a method like https://stanford.edu/~boyd/papers/pdf/qcqp.pdf shall be used.

**Comment by xinyueshen** [comment]
@kosmitive Thanks for the details. One of these errors reports that Q is indefinite, but if it is positive definite, then DCCP should be able to work. For example the following code works.
```
import dccp
import cvxpy as cp
import numpy as np
np.random.seed(0)

t_var = cp.Variable(1)
x = cp.Variable(2)
Q = cp.Parameter((2,2), PSD=True)
obj = cp.quad_form(x, Q)
reg = cp.sum_squares(x)
constraints = [x >= 0, x <= 10, obj == t_var]
cost = reg - t_var
prob = cp.Problem(cp.Minimize(cost), constraints)

Q.value = 1.5 * np.eye(2)
prob.solve(method="dccp", tau=1.0)
```

**Comment by kosmitive** [comment]
@xinyueshen Q is approximated as the sum of outer product of vectors and the eigenvalues of Q were to small, due to that the error appeared. I quickly solved this by adding a shrinkage matrix to Q to pull the eigenvalues into the positive regime. Now the indefinite error vanishes. Additionally I installed the github repo version instead of the pip version of dccp. This got me through the problem, but I had to finally add tau=1.0 to the solve method in order to obtain the solution. It works now, thanks. :)
